internal/validation/validation.go

```diff
@@ -166,6 +170,93 @@ func validateTypeRestrictions(typesys *typesystem.TypeSystem, tk *openfgav1.Tupl
 	return fmt.Errorf("type '%s' is not an allowed type restriction for '%s#%s'", userType, objectType, tk.GetRelation())
 }
 
+// validateCondition enforces conditions on a relationship tuple
+func validateCondition(typesys *typesystem.TypeSystem, tk *openfgav1.TupleKey) error {
+	objectType := tuple.GetType(tk.Object)
+	userType := tuple.GetType(tk.User)
+	userRelation := tuple.GetRelation(tk.User)
+
+	typeRestrictions, err := typesys.GetDirectlyRelatedUserTypes(objectType, tk.Relation)
+	if err != nil {
+		return err
+	}
+
+	if tk.Condition == nil {
```

Comment:
I think this whole block of code reads a little easier with this alternative implementation. The logic with the negation on the one variable and the non-negation on the other is hard to follow.

```go
if tk.Condition == nil {
	for _, userset := range typeRestrictions {
		if userset.Condition != "" {
			continue
		}

		if userset.Type != userType {
			continue
		}

		if userset.GetRelationOrWildcard() != nil {
			if userset.GetRelation() != "" && userset.GetRelation() != userRelation {
				continue
			}

			if userset.GetWildcard() != nil && !tuple.IsTypedWildcard(tk.User) {
				continue
			}
		}

		return nil
	}

	return &tuple.InvalidConditionalTupleError{
		Cause: fmt.Errorf("condition is missing"), TupleKey: tk,
	}

}
```

Reply:
The idea with this code is to quickly skip over type restrictions that do not define a condition and otherwise require the type restriction to match.

If no such type restrictions are found we return the error, otherwise the first match we find we return. It avoids continuing to cycle through directly related types, and I think it reads a little more straightforward.